enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+
+In the above diagram, each network is assigned a unique conntrack zone and conntrack mark. These are required in order to be able to handle overlapping networks egressing into the same VRF and SNAT’ing to the host IP. More details in the next section.
+
+
+### Shared Gateway Mode
+
+
+#### Pod Egress
+
+On pod egress, the respective GR of that network will handle doing the SNAT to a unique masquerade subnet IP assigned to this network. For example, in the above diagram packets leaving GR-default would be SNAT’ed to 169.254.169.5 in zone 64005. The packet will then enter br-ex, where flows in br-ex will match this packet, and then SNAT the packet to the node IP in zone 0, and apply its CT mark of 5. Finally, the packet will be recirculated back to table 0, where the packet will be CT marked with 1 in zone 64000, and sent out of the physical interface. In OVN-Kubernetes we use zone 64000 to track things from OVN or the host and additionally, we mark packets from OVN with a CT Mark of 1 and packets from the host with 2. Pseudo openflow rules would look like this (assuming node IP of 172.18.0.3):
+
+pod-->GR(snat, 169.254.169.5, zone 64005)->br-ex(snat, 172.18.0.3, zone 0, mark 5, table=0) -->recirc table0 (commit zone 64000, mark 1) -->eth0
+
+The above design will accommodate for overlapping networks with overlapping ports. The worst case scenario is if two networks share the same address space, and two pods with identical IPs are trying to connect externally using the same source and destination port. Although unlikely, we have to plan for this type of scenario. When each pod tries to send a packet through their respective GR, SNAT’ing to the unique GR masquerade IP differentiates the conntrack entries. Now, when the final SNAT occurs in br-ex with zone 0, they can be determined as different connections via source IP, and when SNAT’ing to host IP, conntrack will detect a collision using the same layer 4 port, and choose a different port to use.
+
+When reply traffic comes back into the cluster, we must now submit the packet to conntrack to find which network this traffic belongs to. The packet is always first sent into zone 64000, where it is determined whether or not this packet belonged to OVN (CT mark of 1) or the host. Once identified by CT mark as OVN traffic, the packet will then be unSNAT’ed in zone 0 via br-ex rules and the CT mark restored of which network it belonged to. Finally we can send the packet to the correct GR via the right patch port, by matching on the restored CT Mark. From there, OVN will handle unSNAT’ing the masquerade IP and forward the packet to the original pod.
+
+Also to support kubevirt live migration the GR LRP will have an extra address with the configured gateway for the layer2 subnet so the gateway IP is independent of the node where the VM is running on, after live migration OVN should send and GARP for VMs to cleanup its ARP tables since the gateway IP has different mac now.
```

Comment:
> @qinqon question for you; we identify the migration has completed via the launcher pod annotations right ?
Yes, but... aren't we missing a feature at OVN to explicitly send GARPs ? the only feature that we have there is is to increase the GARP frequency.

> 
> Another question ... does it make sense to define the GW IP on the API ? This could be valuable for localnet for instance (where the GW sits outside the k8s cluster). An alternative would be to default to a well known address - .1 / .254 .
> 
We should have a sane default (.1) and also add a knob to to customize it.

> Anyway, we can start with a well known value, then add it to the API if needed, maintaining the "well known value" as the default.

Reply:
> Yes, but... aren't we missing a feature at OVN to explicitly send GARPs ? the only feature that we have there is is to increase the GARP frequency.

Would you remind me how this GARP is different from the GARP libvirt issues once the migration is complete ? 

Then, the semantics of an OVN initiated GARP elude me. How would OVN know it must send a GARP reacting to a migration completing ? It doesn't care care about migrations / that kind of semantics .

Reply:
> > > Yes, but... aren't we missing a feature at OVN to explicitly send GARPs ? the only feature that we have there is is to increase the GARP frequency.
> > 
> > 
> > Would you remind me how this GARP is different from the GARP libvirt issues once the migration is complete
> 
> libvirt GARP is from the VM to the world, the one we need is from OVN to the world so VM's arp table is updated.

I miss some keywords here from a separate discussion we've had; this GARP's purpose is for the VM to update it's gateway . 

Thanks for explaining.

> 
> > Then, the semantics of an OVN initiated GARP elude me. How would OVN know it must send a GARP reacting to a migration completing ? It doesn't care care about migrations / that kind of semantics .
> 
> Looks like ovn-k is already sending GARPs so we don't need OVN to do so https://github.com/ovn-org/ovn-kubernetes/blob/f5d4dfb6acef57d9a4cad95fedd672dfe4112f9a/go-controller/pkg/node/linkmanager/link_network_manager.go#L193
> 
> ovn-k can just detect live migration finish and trigger it.

Cool! Makes sense to me.